kratos/utilities/sparse_matrix_multiplication_utility.h

```diff
@@ -285,92 +276,75 @@ class SparseMatrixMultiplicationUtility
+                row_width += index1_b[a_col + 1] - index1_b[a_col];
             }
+            rTLS.my_max = std::max(rTLS.my_max, row_width);
+            return rTLS.my_max;
+        });
 
-            #pragma omp critical
-            max_row_width = std::max(max_row_width, my_max);
+        int nthreads = 1;
+        if constexpr(UsesOpenMP) {
+            nthreads = omp_get_max_threads();
         }
 
-    #ifdef _OPENMP
-        const int nthreads = omp_get_max_threads();
-    #else
-        const int nthreads = 1;
-    #endif
-
-        std::vector< std::vector<IndexType> > tmp_col(nthreads);
-        std::vector< std::vector<ValueType> > tmp_val(nthreads);
+        std::vector<std::vector<IndexType>> tmp_col(nthreads);
+        std::vector<std::vector<ValueType>> tmp_val(nthreads);
 
         for(int i = 0; i < nthreads; ++i) {
             tmp_col[i].resize(3 * max_row_width);
             tmp_val[i].resize(2 * max_row_width);
         }
 
-        // We create the c_ptr auxiliar variable
+        // We create the c_ptr auxiliary variable
         IndexType* c_ptr = new IndexType[nrows + 1];
         c_ptr[0] = 0;
 
-        #pragma omp parallel
-        {
-        #ifdef _OPENMP
-            const int tid = omp_get_thread_num();
-        #else
-            const int tid = 0;
-        #endif
+        IndexPartition<IndexType>(nrows).for_each([&](IndexType i) {
+            const IndexType row_beg = index1_a[i];
+            const IndexType row_end = index1_a[i+1];
 
+            int tid = 0;
+            if constexpr(UsesOpenMP) {
+                tid = omp_get_thread_num();
+            }
             IndexType* t_col = &tmp_col[tid][0];
 
-            #pragma omp for
-            for(int i = 0; i < static_cast<int>(nrows); ++i) {
-                const IndexType row_beg = index1_a[i];
-                const IndexType row_end = index1_a[i+1];
-
-                c_ptr[i+1] = ProdRowWidth( index2_a + row_beg, index2_a + row_end, index1_b, index2_b, t_col, t_col + max_row_width, t_col + 2 * max_row_width );
-            }
-        }
+            c_ptr[i+1] = ProdRowWidth( index2_a + row_beg, index2_a + row_end, index1_b, index2_b, t_col, t_col + max_row_width, t_col + 2 * max_row_width );
+        });
 
         // We initialize the sparse matrix
         std::partial_sum(c_ptr, c_ptr + nrows + 1, c_ptr);
         const SizeType nonzero_values = c_ptr[nrows];
         IndexType* aux_index2_c = new IndexType[nonzero_values];
         ValueType* aux_val_c = new ValueType[nonzero_values];
 
-        #pragma omp parallel
-        {
-        #ifdef _OPENMP
-            const int tid = omp_get_thread_num();
-        #else
-            const int tid = 0;
-        #endif
+        IndexPartition<IndexType>(nrows).for_each([&](IndexType i) {
+            const IndexType row_beg = index1_a[i];
+            const IndexType row_end = index1_a[i+1];
+
+            int tid = 0;
+            if constexpr(UsesOpenMP) {
```

Comment:
same

Reply:
It is already used in the current code, just refactoring